glib/src/param_spec.rs

```diff
@@ -384,6 +384,60 @@ macro_rules! define_param_spec_numeric {
     };
 }
 
+macro_rules! define_builder {
+    ($rust_type:ident, $builder_type:ident, $($field_id:ident: $field_ty:ty $(= $field_expr:expr)?,)*) => {
+        #[derive(Default)]
+        pub struct $builder_type<'a> {
+            name: &'a str,
+            nick: Option<&'a str>,
+            blurb: Option<&'a str>,
+            $($field_id: Option<$field_ty>),*
+        }
+        impl<'a> $builder_type<'a> {
+            pub fn new(name: &'a str) -> Self {
+                Self {
+                    name,
+                    ..Default::default()
+                }
+            }
+            /// Default: `self.name`
+            pub fn nick(mut self, nick: &'a str) -> Self {
+                self.nick = Some(nick);
+                self
+            }
+            /// Default: `self.name`
+            pub fn blurb(mut self, blurb: &'a str) -> Self {
+                self.blurb = Some(blurb);
+                self
+            }
+
+            $(
+            $(#[doc = concat!("Default: `", stringify!($field_expr), "`")])?
+            pub fn $field_id(mut self, value: $field_ty) -> Self {
+                self.$field_id = Some(value);
+                self
+            }
+            )*
+
+            pub fn build(self) -> ParamSpec {
+                $rust_type::new(self.name, self.nick.unwrap_or(self.name), self.blurb.unwrap_or(self.name),
+                    $(self.$field_id
+                        $(.or(Some($field_expr.into())))?
+                        .expect(concat!("missing `", stringify!($field_id), "` in `", stringify!($builder_type), "`"))
```

Comment:
Basically, this `expect()` should be gone and the corresponding parameters should be part of the `new()` / `builder()` function.

Reply:
I mean, we can catch this easily at compile time so why only handle it at runtime...

Reply:
But that's why I wanted to have Builders: to have a common interface for creating a `ParamSpec`.

I need to map `#[prop(...build(pameter = value)]` to
```
Builder::new(name).parameter(value)
```

This mapping also makes the `Props` macro compatible with external `ParamSpec`s types (if they have a builder).
Ff the macro becomes specific for each `ParamSpec*::new` we can't have compatibility with external `ParamSpec`s.

Also, It's not unusual for a `Builder` to panic. The builders in `gtk-rs` behave in the same thing.

Reply:
Maybe there could be a separate constructor that is only used by the macro? For normal usage you really don't want to defer things to runtime that can easily be caught at compile time.

Reply:
Idea!
What if I add `#[must_use]` on the methods of the required parameters?

Reply:
That doesn't do what you think it does :)

Reply:
And that can't panic unless the builder was misgenerated.

Reply:
Yes, you are both right. That was a bad example for my argumentation. gtk-rs widgets always have a default value for each parameter, so gtk-rs builders aren't really comparable to these ParamSpecs builders.

Reply:
So, I've decided to make the builders safe by adding the required parameters in the `new` and `builder` functions.

The `Props` macro API will then "directly" call the builder (the first required argument, `name`, will be automatically passed by the `Props` macro. The other required arguments stay the same):
```rust
#[prop(..., builder(required_params).minimum(10))]
```